hr_attendance_schedule/models/resource_calendar_attendance.py

```diff
@@ -0,0 +1,63 @@
+# -*- coding: utf-8 -*-
+from datetime import datetime, timedelta
+
+from odoo import fields, models
+from odoo.tools.misc import DEFAULT_SERVER_DATETIME_FORMAT, DEFAULT_SERVER_DATE_FORMAT
+
+
+class ResourceCalendarAttendance(models.Model):
+    _inherit = 'resource.calendar.attendance'
+
+    grace_before_checkin = fields.Float(help="Grace time before checkin in minutes")
+    grace_after_checkin = fields.Float(help="Grace time after checkin in minutes")
+    grace_before_checkout = fields.Float(help="Grace time before checkout in minutes")
+    grace_after_checkout = fields.Float(help="Grace time after checkout in minutes")
+
+    def get_action_date(self, in_out, employee_id):
+        self._init_datetimes()
+        if self._is_now_allowed(employee_id):
+            matching_date = self._get_matching_date(in_out)
+            if matching_date:
+                offset = matching_date - (self._start_datetime if in_out == 'in' else self._end_datetime)
+                return self._now - offset, False
+        return None, True
+
+    def _init_datetimes(self):
+        self._now = datetime.strptime(fields.Datetime.now(), DEFAULT_SERVER_DATETIME_FORMAT)
```

Comment:
how is this different from `datetime.datetime.utcnow()`?

And if you ever really need to parse a datetime in an odoo context, use https://github.com/odoo/odoo/blob/10.0/odoo/fields.py#L1541

Reply:
I'm a bit concerned about the instance member assignments. Any reason not to pass those values through the call stack in the context?

Reply:
It's so I can mock the current date and time: https://github.com/OCA/hr/pull/365/files#diff-b98e1f7cba7765407b9bab8dc4403ab4R151 

Nice!

You mean like `self.env.context['_now'] = xxx`? If so, that's interresting; what are the advantages?

Reply:
the advantage is that you don't introduce race conditions: All threads (not processes) share the same registry, and if you start relying on values of an instance, you might read a value that some other thread wrote. Very theoretical to nearly impossible, but still worth looking into imho.

Reply:
The context is a frozendict. So I need to make a copy, update it, then use it with self.with_env... That's getting messy. I feel I'm missing something. Do you have a good and clean example to illustrate how to do this properly?

Reply:
`.with_context` is a quite standard pattern in Odoo to pass on stuff through the call stack. You're right, the copying is part of this process, and under the hood this actually causes Odoo to read your record again, because the results can differ per context (we just had a nice conversation about that in https://github.com/OCA/server-tools/pull/969). That indeed might be needlessly inefficient in your case. I pass around some mutable type if I want to split code in functions that shares a set of variables. Usually, I use `namedtuple`s as something like a `struct` equivalent in Python: https://github.com/OCA/server-tools/pull/988/files#diff-f8576cd00057f1477095100637a6bf7aR14

Reply:
I did some modifications and found that my code runs in the order of 50% slower when I use .with_context().

Implying I modified the code so attendance_data is actually passed around as a method argument and I never actually fetch `context['attendance_data']`, consider this:

```
    def get_action_date(self, in_out, employee_id):
        attendance_data = self._create_attendance_data(in_out, employee_id)
        return self._get_action_date(attendance_data)

    def _create_attendance_data(self, in_out, employee_id):
        AttendanceData = namedtuple('AttendanceData', 'in_out employee_id now start_datetime end_datetime')
        now = fields.Datetime.from_string(fields.Datetime.now())
        d_day = self._create_datetime_from_dayofweek()
        start_datetime = d_day + timedelta(hours=self.hour_from)
        end_datetime = d_day + timedelta(hours=self.hour_to)
        if end_datetime <= start_datetime:
            end_datetime += timedelta(days=1)
        return AttendanceData(in_out, employee_id, now, start_datetime, end_datetime)
```
If I call `self._get_action_date` this way instead:

```
self.with_context(attendance_data=attendance_data)._get_action_date(attendance_data)
```

Code runs 50% slower.

Also, if I use `with_context(attendance_data={})`, it does not slow down. So it has something to do with the somewhat heavy attendance_data object. This is no big deal here, but there's no doubt it will impact my future developments. Any thoughts?